### PR TITLE
[quant] Add FixedQParamsFakeQuantize module

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45538 [quant] Add FixedQParamsFakeQuantize module**

Summary:
This is used to simulate fake quantize operation for ops with fixed quantization parameters
e.g. hardsigmoid

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24004795](https://our.internmc.facebook.com/intern/diff/D24004795)

### CODE REVIEW

File: torch/quantization/fake_quantize.py [hunk 1]
```diff
@@ -2,7 +2,43 @@
...
+    def forward(self, x):
+        pass
+
+    @abstractmethod
+    def calculate_qparams(self, **kwargs):
+        pass
+
+    with_args = classmethod(_with_args)
+
+# TODO: inherit from FakeQuantizeBase
```
Comment (vkuzo) [inline torch/quantization/fake_quantize.py:41]: thoughts on an ETA for this? Would be great if the temporary patch did not live in master for long.

File: torch/quantization/fake_quantize.py [hunk 2]
```diff
@@ -61,6 +97,10 @@ def __init__(self, observer=MovingAverageMinMaxObserver, quant_min=0, quant_max=
         self.qscheme = self.activation_post_process.qscheme
         self.ch_axis = self.activation_post_process.ch_axis \
             if hasattr(self.activation_post_process, 'ch_axis') else -1
+        assert _is_per_channel(self.qscheme) or \
+            _is_per_tensor(self.qscheme), \
+            'Only per channel and per tensor quantization are supported in fake quantize' + \
+            ' got qscheme: ' + str(self.qscheme)
 
     @torch.jit.export
```
Comment (vkuzo) [inline torch/quantization/fake_quantize.py:105]: should the options to enable/disable fake_quant/observer go on the base class too?

File: torch/quantization/fake_quantize.py [hunk 3]
```diff
@@ -137,10 +177,51 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
...
+        return self.scale, self.zero_point
+
 default_fake_quant = FakeQuantize.with_args(observer=MovingAverageMinMaxObserver, quant_min=0, quant_max=255,
                                             dtype=torch.quint8, qscheme=torch.per_tensor_affine, reduce_range=True)
 default_weight_fake_quant = FakeQuantize.with_args(observer=MovingAverageMinMaxObserver, quant_min=-128, quant_max=127,
                                                    dtype=torch.qint8, qscheme=torch.per_tensor_symmetric, reduce_range=False)
+default_symmetric_fixed_qparams_fake_quant = FixedQParamsFakeQuantize.with_args(
+    scale=2.0 / 256.0, zero_point=128, dtype=torch.quint8)
+default_affine_fixed_qparams_fake_quant = FixedQParamsFakeQuantize.with_args(
+    scale=1.0 / 256.0, zero_point=0, dtype=torch.quint8, quant_min=-128, quant_max=127)
```
Comment (vkuzo) [inline torch/quantization/fake_quantize.py:224]: are these tied to a specific function? If so, maybe we could include the function name in the var name, since the range depends on the input function?

**Review by vkuzo — APPROVED** [review]

**Comment by jerryzh168** [comment]
this is not needed anymore, we plan to observe the output for all the ops that inherits quantization parameters (e.g. average pool) and ops that currently have fixed quantization parameters (e.g. sigmoid).

File: torch/quantization/fake_quantize.py (hunk 1 above)
Comment (jerryzh168) [reply torch/quantization/fake_quantize.py:41]: will do this in 1-2 month

File: torch/quantization/fake_quantize.py (hunk 2 above)
Comment (jerryzh168) [reply torch/quantization/fake_quantize.py:105]: I think ideally these should be removed

Comment (jerryzh168) [reply torch/quantization/fake_quantize.py:105]: but as a middle step, it should live in base class as well.

File: torch/quantization/fake_quantize.py (hunk 3 above)
Comment (jerryzh168) [reply torch/quantization/fake_quantize.py:224]: there are multiple functions that might use these, the mapping will live in a different place.
